# 16 — Physical categorical hierarchies

This notebook presents the frozen-fold extraction, source, quality and waterpoint hierarchy evidence. The accepted model recipe stays fixed, the labelled local test remains unopened, and no oversampling or competition prediction participates.

In [1]:
from pathlib import Path
import pandas as pd

project_dir = next(
    path
    for path in (Path.cwd(), *Path.cwd().parents)
    if (path / 'stage-1-pump-it-up').exists()
)
runtime_dir = project_dir / '.runtime' / 'physical-hierarchy-screen'
families = {
    family: pd.read_csv(runtime_dir / f'{family}-summary.csv', index_col='policy')
    for family in ('extraction', 'source', 'quality', 'waterpoint')
}
hybrids = pd.read_csv(runtime_dir / 'component-hybrids.csv', index_col='hybrid')
print(f"Policies: {sum(len(frame) for frame in families.values())}; refitted challengers: 20")
print(f"Component hybrids: {len(hybrids)}")

Policies: 24; refitted challengers: 20
Component hybrids: 6


## Audit boundary

Every hierarchy is deterministic, complete, and covered by the competition frame: extraction 18→13→7 levels, source 10→7→3, quality 8→6, and waterpoint 7→6. This is therefore a granularity and redundancy screen rather than an imputation exercise.

In [2]:
leader_rows = []
for family, frame in families.items():
    leader = frame.iloc[0]
    leader_rows.append({
        'family': family,
        'policy': frame.index[0],
        'mean_accuracy': leader['mean_accuracy'],
        'accuracy_change': leader['accuracy_change'],
        'fold_wins': int(leader['fold_wins']),
        'passes_gate': bool(leader['passes_gate']),
    })
pd.DataFrame(leader_rows).set_index('family').style.format({
    'mean_accuracy': '{:.3%}',
    'accuracy_change': '{:+.3%}',
})

,policy,mean_accuracy,accuracy_change,fold_wins,passes_gate
family,,,,,
extraction,baseline,81.625%,+0.000%,0,False
source,source_plus_class,81.635%,+0.011%,2,False
quality,baseline,81.625%,+0.000%,0,False
waterpoint,both_levels,81.627%,+0.002%,2,False


## Extraction and source hierarchies

In [3]:
columns = ['label', 'transformed_feature_change', 'mean_accuracy', 'accuracy_change', 'fold_wins', 'worst_fold_change', 'repair_recall', 'passes_gate', 'parsimony_candidate']
display(families['extraction'][columns].style.format({
    'mean_accuracy': '{:.3%}', 'accuracy_change': '{:+.3%}',
    'worst_fold_change': '{:+.3%}', 'repair_recall': '{:.3%}',
}))
display(families['source'][columns].style.format({
    'mean_accuracy': '{:.3%}', 'accuracy_change': '{:+.3%}',
    'worst_fold_change': '{:+.3%}', 'repair_recall': '{:.3%}',
}))

,label,transformed_feature_change,mean_accuracy,accuracy_change,fold_wins,worst_fold_change,repair_recall,passes_gate,parsimony_candidate
policy,,,,,,,,,
baseline,Granular extraction type,0,81.625%,+0.000%,0,+0.000%,34.859%,False,False
group_only,Extraction group only,-4,81.593%,-0.032%,2,-0.168%,35.032%,False,True
type_plus_class,Extraction type plus class,7,81.576%,-0.048%,1,-0.105%,35.004%,False,False
all_levels,All extraction levels,20,81.576%,-0.048%,1,-0.168%,34.888%,False,False
type_plus_group,Extraction type plus group,13,81.574%,-0.051%,1,-0.221%,34.772%,False,False
group_plus_class,Extraction group plus class,3,81.540%,-0.084%,2,-0.168%,34.627%,False,False
class_only,Extraction class only,-10,81.402%,-0.223%,1,-0.347%,34.656%,False,False
none,No extraction hierarchy,-17,81.174%,-0.450%,0,-0.673%,34.540%,False,False


,label,transformed_feature_change,mean_accuracy,accuracy_change,fold_wins,worst_fold_change,repair_recall,passes_gate,parsimony_candidate
policy,,,,,,,,,
source_plus_class,Source plus class,3,81.635%,+0.011%,2,-0.105%,34.859%,False,False
baseline,Granular source,0,81.625%,+0.000%,0,+0.000%,34.859%,False,False
type_only,Source type only,-3,81.572%,-0.053%,2,-0.147%,34.946%,False,False
type_plus_class,Source type plus class,0,81.536%,-0.088%,0,-0.137%,34.888%,False,False
all_levels,All source levels,10,81.534%,-0.090%,0,-0.231%,34.830%,False,False
source_plus_type,Source plus type,7,81.505%,-0.120%,1,-0.263%,34.541%,False,False
class_only,Source class only,-7,81.479%,-0.145%,1,-0.295%,34.280%,False,False
none,No source hierarchy,-10,81.353%,-0.271%,0,-0.347%,34.338%,False,False


## Quality and waterpoint hierarchies

In [4]:
display(families['quality'][columns].style.format({
    'mean_accuracy': '{:.3%}', 'accuracy_change': '{:+.3%}',
    'worst_fold_change': '{:+.3%}', 'repair_recall': '{:.3%}',
}))
display(families['waterpoint'][columns].style.format({
    'mean_accuracy': '{:.3%}', 'accuracy_change': '{:+.3%}',
    'worst_fold_change': '{:+.3%}', 'repair_recall': '{:.3%}',
}))

,label,transformed_feature_change,mean_accuracy,accuracy_change,fold_wins,worst_fold_change,repair_recall,passes_gate,parsimony_candidate
policy,,,,,,,,,
baseline,Granular water quality,0,81.625%,+0.000%,0,+0.000%,34.859%,False,False
both_levels,Water quality plus group,6,81.561%,-0.063%,1,-0.221%,34.772%,False,False
group_only,Quality group only,-2,81.538%,-0.086%,1,-0.168%,34.395%,False,False
none,No quality hierarchy,-8,81.399%,-0.225%,0,-0.421%,34.569%,False,False


,label,transformed_feature_change,mean_accuracy,accuracy_change,fold_wins,worst_fold_change,repair_recall,passes_gate,parsimony_candidate
policy,,,,,,,,,
both_levels,Waterpoint type plus group,6,81.627%,+0.002%,2,-0.084%,35.032%,False,False
baseline,Granular waterpoint type,0,81.625%,+0.000%,0,+0.000%,34.859%,False,False
group_only,Waterpoint group only,-1,81.383%,-0.242%,0,-0.368%,34.222%,False,False
none,No waterpoint hierarchy,-7,81.048%,-0.577%,0,-0.747%,34.309%,False,False


## Bounded component crosses

Only policy-specific components whose standalone component mean improved are crossed. These use saved aligned out-of-fold probabilities and require no refitting.

In [5]:
hybrids[['label', 'mean_accuracy', 'accuracy_change', 'fold_wins', 'worst_fold_change', 'repair_recall', 'passes_gate']].style.format({
    'mean_accuracy': '{:.3%}',
    'accuracy_change': '{:+.3%}',
    'worst_fold_change': '{:+.3%}',
    'repair_recall': '{:.3%}',
})

,label,mean_accuracy,accuracy_change,fold_wins,worst_fold_change,repair_recall,passes_gate
hybrid,,,,,,,
waterpoint_xgb_source_rf,Waterpoint-both XGBoost + source-class Random Forest,81.629%,+0.004%,2,-0.095%,35.004%,False
waterpoint_xgb_baseline_rf,Waterpoint-both XGBoost + baseline Random Forest,81.582%,-0.042%,1,-0.116%,34.917%,False
baseline_xgb_source_rf,Baseline XGBoost + source-class Random Forest,81.580%,-0.044%,1,-0.137%,34.714%,False
extraction_xgb_baseline_rf,Extraction-group XGBoost + baseline Random Forest,81.574%,-0.051%,1,-0.147%,34.656%,False
quality_xgb_baseline_rf,Quality-group XGBoost + baseline Random Forest,81.561%,-0.063%,0,-0.126%,34.453%,False
baseline_xgb_extraction_rf,Baseline XGBoost + extraction-group Random Forest,81.559%,-0.065%,2,-0.242%,34.830%,False


## Decision

Retain all four granular features in the accepted competition recipe. No policy or component hybrid passes the promotion gate, so none triggers an LGA-disjoint sensitivity run. Extraction group-only remains a named parsimony candidate: five fewer encoded columns for a 0.032-point mean cost. Stop deterministic hierarchy permutation on these folds.